### PR TITLE
GNPSExport UTIL commit

### PR DESCRIPTION
**Pull request by abisarvepalli**

### CODE REVIEW

**Comment by hroest** [comment]
thansk for your contribution, can you please add a test here: https://github.com/OpenMS/OpenMS/blob/develop/src/tests/topp/CMakeLists.txt (see https://github.com/OpenMS/OpenMS/wiki/Write-tests#topp_tests). Also please fix the following issue:

http://cdash.openms.de/testDetails.php?test=51151534&build=280111

```
/home/travis/build/OpenMS/OpenMS/src/utils/GNPSExport.cpp:66:  Missing space before {  [whitespace/braces] [5]
/home/travis/build/OpenMS/OpenMS/src/utils/GNPSExport.cpp:249:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Done processing /home/travis/build/OpenMS/OpenMS/src/utils/GNPSExport.cpp
Total errors found: 2
```

and merge with the `develop` branch again.

**Comment by abisarvepalli** [comment]
image
I have tests but the module keeps failing the TOPPWRITECTD_GNPSExport test. Is there a specific debug for this?

**Comment by hroest** [comment]
yes, you can use `ctest -R GNPSExport` on your local machine

File: src/openms/source/APPLICATIONS/ToolHandler.cpp [hunk 1]
```diff
@@ -74,6 +74,7 @@ namespace OpenMS
     tools_map["FileFilter"] = Internal::ToolDescription("FileFilter", "File Handling");
     tools_map["FileInfo"] = Internal::ToolDescription("FileInfo", "File Handling");
     tools_map["FileMerger"] = Internal::ToolDescription("FileMerger", "File Handling");
+    tools_map["GNPSExport"] = Internal::ToolDescription("GNPSExport", "File Handling");
```
Comment (hroest) [inline src/openms/source/APPLICATIONS/ToolHandler.cpp]: I have added the tool to the ToolHandler, that should remove the warning

File: src/tests/topp/CMakeLists.txt [hunk 2]
```diff
@@ -1094,15 +1105,15 @@ ${TOPP_BIN_PATH}/OpenSwathAssayGenerator -test -in ${DATA_DIR_TOPP}/OpenSwathAss
...
-${TOPP_BIN_PATH}/OpenSwathDecoyGenerator -test -in ${DATA_DIR_TOPP}/OpenSwathDecoyGenerator_input_2.TraML -out OpenSwathDecoyGenerator_2.TraML.tmp -out_type TraML -method pseudo-reverse -product_mz_threshold 0.8 -switchKR false)
+${TOPP_BIN_PATH}/OpenSwathDecoyGenerator -test -in ${DATA_DIR_TOPP}/OpenSwathDecoyGenerator_input_2.TraML -out OpenSwathDecoyGenerator_2.TraML.tmp -out_type TraML -method pseudo-reverse -product_mz_threshold 0.8)
   add_test("TOPP_OpenSwathDecoyGenerator_test_2_out1" ${DIFF} -in1 OpenSwathDecoyGenerator_2.TraML.tmp -in2 ${DATA_DIR_TOPP}/OpenSwathDecoyGenerator_output_2.TraML)
   set_tests_properties("TOPP_OpenSwathDecoyGenerator_test_2_out1" PROPERTIES DEPENDS "TOPP_OpenSwathDecoyGenerator_test_2")
   add_test("TOPP_OpenSwathDecoyGenerator_test_3"
-${TOPP_BIN_PATH}/OpenSwathDecoyGenerator -test -in ${DATA_DIR_TOPP}/OpenSwathDecoyGenerator_input_3.TraML -out OpenSwathDecoyGenerator_3.TraML.tmp -out_type TraML -method pseudo-reverse -separate -switchKR false)
+${TOPP_BIN_PATH}/OpenSwathDecoyGenerator -test -in ${DATA_DIR_TOPP}/OpenSwathDecoyGenerator_input_3.TraML -out OpenSwathDecoyGenerator_3.TraML.tmp -out_type TraML -method pseudo-reverse -separate)
```
Comment (hroest) [inline src/tests/topp/CMakeLists.txt]: can you undo these changes?

File: src/tests/topp/CMakeLists.txt [hunk 3]
```diff
@@ -300,25 +300,17 @@ add_test("TOPP_FeatureFinderMultiplex_2_out1" ${DIFF} -whitelist "id=" "href=" -
...
+add_test("TOPP_FeatureFinderMultiplex_4_out2" ${DIFF} -whitelist "id=" "href=" -in1 FeatureFinderMultiplex_8.tmp -in2 ${DATA_DIR_TOPP}/FeatureFinderMultiplex_3_output.consensusXML )
 set_tests_properties("TOPP_FeatureFinderMultiplex_4_out2" PROPERTIES DEPENDS "TOPP_FeatureFinderMultiplex_4")
 
-add_test("TOPP_FeatureFinderMultiplex_5" ${TOPP_BIN_PATH}/FeatureFinderMultiplex -test -in ${DATA_DIR_TOPP}/FeatureFinderMultiplex_5_input.mzML -ini ${DATA_DIR_TOPP}/FeatureFinderMultiplex_5_parameters.ini -out FeatureFinderMultiplex_9.tmp -out_multiplets FeatureFinderMultiplex_10.tmp)
-add_test("TOPP_FeatureFinderMultiplex_5_out1" ${DIFF} -whitelist "id=" "href=" -in1 FeatureFinderMultiplex_9.tmp -in2 ${DATA_DIR_TOPP}/FeatureFinderMultiplex_5_output.featureXML )
-set_tests_properties("TOPP_FeatureFinderMultiplex_5_out1" PROPERTIES DEPENDS "TOPP_FeatureFinderMultiplex_5")
-add_test("TOPP_FeatureFinderMultiplex_5_out2" ${DIFF} -whitelist "id=" "href=" -in1 FeatureFinderMultiplex_10.tmp -in2 ${DATA_DIR_TOPP}/FeatureFinderMultiplex_5_output.consensusXML )
-set_tests_properties("TOPP_FeatureFinderMultiplex_5_out2" PROPERTIES DEPENDS "TOPP_FeatureFinderMultiplex_5")
-
```
Comment (hroest) [inline src/tests/topp/CMakeLists.txt]: can you undo these changes?

File: src/tests/topp/CMakeLists.txt [hunk 4]
```diff
@@ -675,6 +667,25 @@ set_tests_properties("TOPP_FileMerger_10_out2" PROPERTIES DEPENDS "TOPP_FileMerg
...
+add_test("TOPP_GNPSExport_2" ${TOPP_BIN_PATH}/GNPSExport -in_cm ${DATA_DIR_TOPP}/GNPSExport_2_inp1.consensusXML -in_mzml ${DATA_DIR_TOPP}/GNPSExport_2_inp2.mzML -out GNPSExport_2_out.tmp -output_type merged_spectra)
+add_test("TOPP_GNPSExport_2_out" ${DIFF} -in1 GNPSExport_2_out.tmp -in2 ${DATA_DIR_TOPP}/GNPSExport_2_out.mgf)
+
+# two files using full_spectra:
+add_test("TOPP_GNPSExport_3" ${TOPP_BIN_PATH}/GNPSExport -in_cm ${DATA_DIR_TOPP}/GNPSExport_3_inp1.consensusXML -in_mzml ${DATA_DIR_TOPP}/GNPSExport_3_inp2.mzML ${DATA_DIR_TOPP}/GNPSExport_3_inp3.mzML -out GNPSExport_3_out.tmp -output_type full_spectra)
+add_test("TOPP_GNPSExport_3_out" ${DIFF} -in1 GNPSExport_3_out.tmp -in2 ${DATA_DIR_TOPP}/GNPSExport_3_out.mgf)
+
+# two files using merged_spectra:
+add_test("TOPP_GNPSExport_4" ${TOPP_BIN_PATH}/GNPSExport -in_cm ${DATA_DIR_TOPP}/GNPSExport_4_inp1.consensusXML -in_mzml ${DATA_DIR_TOPP}/GNPSExport_4_inp2.mzML ${DATA_DIR_TOPP}/GNPSExport_4_inp3.mzML -out GNPSExport_4_out.tmp)
+add_test("TOPP_GNPSExport_4_out" ${DIFF} -in1 GNPSExport_4_out.tmp -in2 ${DATA_DIR_TOPP}/GNPSExport_4_out.mgf)
```
Comment (hroest) [inline src/tests/topp/CMakeLists.txt]: you will need to add dependencies:
```
set_tests_properties("TOPP_GNPSExport_4_out" PROPERTIES DEPENDS "TOPP_GNPSExport_4") 
```
for all your tests since they get executed in parallel and you need to *first* execute the tool before you can test its results

**Comment by hroest** [comment]
can you fix these issues:

http://cdash.openms.de/testDetails.php?test=51334322&build=281960
```
/home/travis/build/OpenMS/OpenMS/src/topp/GNPSExport.cpp:66:  Missing space before {  [whitespace/braces] [5]
/home/travis/build/OpenMS/OpenMS/src/topp/GNPSExport.cpp:250:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Done processing /home/travis/build/OpenMS/OpenMS/src/topp/GNPSExport.cpp
Total errors found: 2
```

also please fix  src/tests/topp/CMakeLists.txt by undoing some of your changes to other parts of the file and adding test properties

**Comment by hroest** [comment]
looks much better, only 2 errors remain: http://cdash.openms.de/viewTest.php?onlyfailed&buildid=282629

however, it seems some input / output got mixed up:

```

Offending lines:			(tab_width = 8, first_column = 1)

in1:  /home/jenkins/ws/openms_pr_ci/a9154670/build/src/tests/topp/GNPSExport_4_out.tmp   (line: 20548, position/column: 25/26)
FILENAME=GNPSExport_4_inp!
                         3.mzML

in2:  /home/jenkins/ws/openms_pr_ci/a9154670/source/src/tests/topp/GNPSExport_4_out.mgf   (line: 20548, position/column: 25/26)
FILENAME=GNPSExport_4_inp!
                         2.mzML


```

File: src/topp/GNPSExport.cpp [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+									ms2_block[curr_spectrum.getMZ()] = curr_spectrum.getIntensity();
+								}
+							}
+						}
+					}
+
+					// SpectraMerger merger;
+					// Param p;
+					// p.setValue("precursor_method:mz_tolerance", prec_mz_tol);
+					// p.setValue("precursor_method:rt_tolerance", prec_rt_tol*2);
```
Comment (oliveralka) [inline src/topp/GNPSExport.cpp]: How do you merge the spectra in your merged spectra section (e.g. small comment)?

Could you please remove unused commented code (e.g.  SpectraMerger).

**Comment by hroest** [comment]
looks like still 2 tests are failing:
```
TOPP_GNPSExport_3_out | Failed | 270ms | Completed (Failed)
TOPP_GNPSExport_4_out | Failed | 190ms | Completed (Failed
```
http://cdash.openms.de/viewTest.php?onlyfailed&buildid=283518

**Comment by hroest** [comment]
regarding the tests: it is great to have tests, but could you please chose a test file < 1 MB (your current file GNPSExport_1_inp2.mzML is 32 MB large). We dont want to add that much data to the OpenMS repositories...

**Comment by abisarvepalli** [comment]
the issue is that it passes on my current os (macos highsierra), but when running it via the pr, it seems to fail.

**Comment by hroest** [comment]
> the issue is that it passes on my current os (macos highsierra), but when running it via the pr, it seems to fail.

yes, it seems that it fails on Windows and Linux but passes on OSX. It is likely that this is a sorting issue in your output or a variable that you forgot to initialize. Note that your tests need to pass on all platforms before we can merge.

File: src/topp/GNPSExport.cpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+		registerDoubleOption_("precursor_rt_tolerance", "<num>", DEF_PRECURSOR_RT_TOLERANCE, "Tolerance rt window for precursor selection", false);
+
+		registerTOPPSubsection_("merged_spectra", "Options for exporting mgf file with merged spectra per feature");
+		registerDoubleOption_("merged_spectra:cos_similarity", "<num>", DEF_COSINE_SIMILARITY, "Cosine similarity threshold for merged_spectra output", false);
+	}
+
+	// the main function is called after all parameters are read
+	ExitCodes main_(int, const char **)
+	{
+		ProgressLogger progress_logger;
```
Comment (hroest) [inline src/topp/GNPSExport.cpp]: could you follow our conventions and use two spaces instead of tabs for indent?

**Comment by abisarvepalli** [comment]
image
it seems that the GNPSExport cpplint test had failed but was logged as it being failed? On another note, all the other tests passed!

**Comment by oliveralka** [comment]
These are the errors for cpplint, pease have a look:
http://cdash.openms.de/testDetails.php?test=51768699&build=283803

File: src/topp/GNPSExport.cpp [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+    }
+
+
+    //-------------------------------------------------------------
+    // calculations
+    //-------------------------------------------------------------
+    progress_logger.startProgress(0, consensus_map.size(), "parsing features and ms2 identifications...");
+    std::stringstream output_stream;
+    Size feature_count = 1;
+    for(Size i = 0; i != consensus_map.size(); ++i) {
```
Comment (hroest) [inline src/topp/GNPSExport.cpp]: can you change the style to
```
for (...)
{
}
```

and follow our https://github.com/OpenMS/OpenMS/wiki/Coding-conventions

**Comment by abisarvepalli** [comment]
woo finally!

**Comment by hroest** [comment]
rebuild jenkins

**Comment by hroest** [comment]
@timosachsenberg @jpfeuffer please remember to squash the merge since otherwise we will have all the large files in the OpenMS history!
